ops/hookcmds/_types.py

```diff
@@ -0,0 +1,384 @@
+
+    status: str
+    since: datetime.datetime
+
+    @classmethod
+    def _from_dict(cls, d: GoalDict) -> Goal:
+        return cls(
+            status=d['status'],
+            since=datetime_from_iso(d['since']),
+        )
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class GoalState:
+    """The units and relations that the model should have, and the status of achieving that."""
+
+    units: dict[str, Goal]
+    # The top key is the endpoint/relation name, the second key is the app/unit name.
+    relations: dict[str, dict[str, Goal]]
+
+    @classmethod
+    def _from_dict(cls, d: GoalStateDict) -> GoalState:
+        units: dict[str, Goal] = {
+            name: Goal._from_dict(unit) for name, unit in d.get('units', {}).items()
+        }
+        relations: dict[str, dict[str, Goal]] = {}
+        for name, relation in d.get('relations', {}).items():
+            goals: dict[str, Goal] = {
+                app_or_unit: Goal._from_dict(data) for app_or_unit, data in relation.items()
+            }
+            relations[name] = goals
+        return cls(units=units, relations=relations)
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class Network:
+    """A Juju space."""
+
+    bind_addresses: Sequence[BindAddress]
+    egress_subnets: Sequence[ipaddress.IPv4Network | ipaddress.IPv6Network]
+    ingress_addresses: Sequence[ipaddress.IPv4Address | ipaddress.IPv6Address]
+
+    @classmethod
+    def _from_dict(cls, d: dict[str, Any]) -> Network:
+        bind: list[BindAddress] = [
+            BindAddress._from_dict(bind_data)
+            for bind_data in cast('list[BindAddressDict]', d['bind-addresses'])
+        ]
+        egress = [ipaddress.ip_network(addr) for addr in d.get('egress-subnets', [])]
+        ingress = [ipaddress.ip_address(addr) for addr in d.get('ingress-addresses', [])]
+        return cls(bind_addresses=bind, egress_subnets=egress, ingress_addresses=ingress)
+
+
+# Note that we intend to merge this with model.py's `Port` in the future, and
+# that does not have `kw_only=True`. That means that we should not use it here,
+# either, so that merging can be backwards compatible.
+@dataclasses.dataclass(frozen=True)
+class Port:
+    """A port that Juju has opened for the charm."""
+
+    protocol: Literal['tcp', 'udp', 'icmp'] | None = 'tcp'
+    """The IP protocol."""
+
+    port: int | None = None
+    """The port number. Will be ``None`` if protocol is ``'icmp'``."""
+
+    to_port: int | None = None
+    """The final port number if this is a range of ports."""
+
+    endpoints: list[str] | None = None
+    """The endpoints this port applies to, ``['*']`` if all endpoints, or ``None`` if unknown."""
+
+
+class RelationModelDict(TypedDict):
+    uuid: str
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class RelationModel:
+    """Details of the model on the remote side of the relation."""
+
+    uuid: uuid.UUID
```

Comment:
I'm a little hesitant to make this `uuid.UUID` type instead of just `str`. Seems like it'd be somewhat likely Juju would upgrade to something Python's uuid module doesn't support, or an xid or something like secrets. What do you think?

Reply:
Changed to `str`.